### PR TITLE
Implement visibility range and dependencies

### LINKED ISSUE #692: Add support for 3D Level of Detail
I'm going to ignore the issue template because this has been confirmed to be implemented in the future.

I'm opening this issue because the conversation arround LOD is scattered by multiple issues and I think it is important to have a common place to discuss the technical details of a feature (and move the discussion of the design to the proposals repository).

### Hierarchical level of detail
Another LOD approach (and not mutually exclusive with the previously mentioned) is the hierarchical level of detail. With this a single mesh can replace multiple smaller ones. It greatly helps reducing the mesh count for the areas far from the camera.
https://docs.unrealengine.com/en-US/Engine/HLOD/index.html

This will be implemented as stated here: https://twitter.com/reduzio/status/1340114705923379202

### Content streaming

Related issue: godotengine/godot#17556

As indicated by @Calinou : "Support for LOD and texture streaming is planned in Godot 4.0"

### PR DESCRIPTION
**Pull request by JFonS**
This commit adds the following properties to GeometryInstance3D: `visibility_range_begin`,
`visibility_range_begin_margin`, `visibility_range_end`, `visibility_range_end_margin`.

Together they define a range in which the GeometryInstance3D will be visible from the camera,
taking hysteresis into account for state changes. A begin or end value of 0 will be ignored,
so the visibility range can be open-ended in both directions.

This commit also adds the `visibility_parent` property to 'Node3D'.
Which defines the visibility parents of the node and its subtree (until
another parent is defined).

Visual instances with a visibility parent will only be visible when the parent, and all of its
ancestors recursively, are hidden because they are closer to the camera than their respective
`visibility_range_begin` thresholds.

Combining visibility ranges and visibility parents users can set-up a quick HLOD system
that shows high detail meshes when close (i.e buildings, trees) and merged low detail meshes
for far away groups (i.e. cities, woods).

This is a simple example "city":

https://user-images.githubusercontent.com/4402304/118841427-99e7b000-b8c8-11eb-9bcd-53b7fe65c1df.mp4

And this is a stress test with 50k of such "cities" (900k meshes in total):

https://user-images.githubusercontent.com/4402304/118842300-5fcade00-b8c9-11eb-9bfe-ba4caeb5d527.mp4

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/692.*

### CODE REVIEW

File: doc/classes/GeometryInstance3D.xml
Comment (clayjohn) [inline doc/classes/GeometryInstance3D.xml:63]: Maybe add another sentence about what hysteresis is used for. Something like "hysteresis is the amount of overlap between two visibility ranges allowing both HLODs to draw at the same time to reduce popping in and out"

File: scene/3d/visual_instance_3d.cpp
Comment (clayjohn) [inline scene/3d/visual_instance_3d.cpp:185]: typo, should be "hysteresis", this is spelled wrong throughout

**Review by clayjohn — COMMENTED** [review]
Looks really good on a quick review. You'll need to do a search and replace to replace all instances of "hysteriesis" with "hysteresis"

**Comment by fire** [comment]
Are you using the LOD metric that is used by mesh optimizer?

**Comment by JFonS** [comment]
> Looks really good on a quick review. You'll need to do a search and replace to replace all instances of "hysteriesis" with "hysteresis"
Oops, will do :P

> Are you using the LOD metric that is used by mesh optimizer?
No, AFAIK the mesh LOD metric uses the edge lengths and this system has to work on all GeometryInstance3D types. Also, a simple distance  check with hysteresis should work just fine for this kind of HLOD.

File: doc/classes/GeometryInstance3D.xml
Comment (JFonS) [reply doc/classes/GeometryInstance3D.xml:63]: Right, I will add something like that :)

**Comment by Calinou** [comment]
Does this visibility range system take the camera FOV into account? I wonder how we can keep using ranges while supporting wider/narrower camera FOVs.

**Review by clayjohn — COMMENTED** [review]
Looks good to me!

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
I am thinking that as much as it pains me, we should probably change it to a system where the parent has a visibility override over the children and not backwards, else usability will be poor.
This also means storing a list of children in the node, which will make editing more difficult and it may require a special editor. I can work on it I suppose, though I am just unhappy.

**Comment by JFonS** [comment]
I updated the PR with the following changes:

* `visibility_range_begin` and `visibility_range_end` are disabled if their value is 0. This means I was able to remove the previous `visibility_range_enabled` property, and visibility checks are internally enabled only when any of the range values is not 0.
* I changed the visibility_parent behavior so that it only shows it's "children" instances when the camera is too close, while before it would show children whenever the instance was outside of the visibility range (both too far and too close).  This means most LOD levels don't have to check for a "range_end" threshold anymore, making the system more user-friendly and faster.

File: scene/3d/visual_instance_3d.cpp
Comment (reduz) [inline scene/3d/visual_instance_3d.cpp:455]: I would most likely do this internally, and not expose the parent range or clearing parent to the user. As in, user should only need to be able to set the children.

File: scene/3d/visual_instance_3d.cpp
Comment (reduz) [inline scene/3d/visual_instance_3d.cpp:229]: should probably use NOTIFICATION_READY to make sure children are properly configured (and call request_ready() afterwards so its called again next time it enters the tree.)

**Review by reduz — CHANGES_REQUESTED** [review]

File: scene/3d/visual_instance_3d.cpp
Comment (JFonS) [reply scene/3d/visual_instance_3d.cpp:455]: What do you mean with "set the children"? Exposing a method that receives a list of nodes and then sets their parent internally?

The issue with that is it hides the fact that nodes can only have one parent, if properly documented it won't be that bad, but I'm still not sure what are the benefits of setting the children instead of setting the parent...

I can see why it's needed in the editor, where you just edit one node and select all of its children, but from the scripting side it's just a matter of iterating your list of children and setting their parent node.

File: scene/3d/visual_instance_3d.cpp
Comment (JFonS) [reply scene/3d/visual_instance_3d.cpp]: I don't think this is needed anymore. Right now, the only requirement is for the node to be inside a tree. If any nodes are added after that, they will automatically inherit the correct visibility parent.

File: scene/3d/visual_instance_3d.cpp
Comment (JFonS) [reply scene/3d/visual_instance_3d.cpp]: Fixed in the latest version of the PR.

**Comment by JFonS** [comment]
Pushed the latest version with a new way to define visibility dependencies and fixes for all the mentioned issues.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
Looks fantastic, really great job with this!

**Comment by JFonS** [comment]
I would wait for #48207 to be merged before merging this one. I will take care of the merge conflicts then

**Comment by JFonS** [comment]
@akien-mga Rebased!

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
